docs/proposals-accepted/202206-block-series-deletion-compactor.md

```diff
@@ -0,0 +1,206 @@
+  int64 max_time = 2;
+  repeated LabelMatcher matchers = 3 [(gogoproto.nullable) = false];
+}
+
+message TombstoneResponse {
+  repeated SeriesIntervals series_intervals = 1;
+}
+
+message SeriesIntervals {
+  int64 series_ref = 1;
+  repeated Interval intervals = 2;
+}
+
+message Interval {
+  int64 start = 1;
+  int64 end = 2;
+}
+```
+
+After getting the query result, the compactor does almost the same thing as the previous option: generate the tombstone file during compaction time.
+
+The drawback for this approach is that we are inventing a new API for this purpose only, which might be a bit of pain in the future.
+
+#### Option 3: Delete Series using Rewrite Logic
+
+The previous two options do the series deletion at compaction time, this option does actual series deletion using block rewrite. For the index scanning, it is still required to choose either option 1 or 2 to avoid unncessary rewrite even if the block doesn't have any matched series.
+
+This is not so efficient as rewrite is a different process than compaction. We need to add this step and make sure rewrite can work together with compaction and no redundant work is done.
+
+Overall, among the 3 options, option 1 is better as it is easier to implement and straightforward.
+
+### Other Considerations
+
+#### Compaction Planner for Tombstone
+
+If series deletion is done at compaction time, then no much thing to change at the planner as we just need to delete series at compaction time. For the ratio calculation, as the tombstone results are pre-calculated it is easy to know the ratio as well. The threshold can be made configurable too.
+
+The exception is for blocks that at max compaction level, no compaction will happen to it and the tombstones will remain forever if the ratio is lower than the deletion threshold. In this case the planner can check the block compaction level first, if it is max compaction level and the tombstone exists for some period of time, then do compaction for that block.
+
+#### Tombstone File Deletion
+
+After one compaction iteration is done, metadata will be synced again and compactor starts doing garbage collection. This time the compactor can sync tombstones again and check is there any matched block for the tombstone file. If no, then the tombstone file can be deleted.
+
+For any new blocks that are compacted by blocks along with some tombstone requests, we want to avoid calculating tombstones on it again as those series are deleted from it already. We propose a new field in the `meta.json` file called `tombstone_applied` to indicate what tombstones have been applied to the block.
+
+Tombstones will skip blocks that have already been applied with its ID.
+
+### Tombstones sharding
+
+Unlike blocks, we cannot shard tombstones simply using its ULIDs because each tombstone is able to match all blocks from the object storage. When compactor sharding is enabled, it still needs to check all the tombstones in the shard to find the matched blocks.
```

Comment:
The tricky thing of only sharding blocks, not tombstones is that we cannot know whether all matched blocks are processed for the given tombstone as each compactor only has information about its own blocks